sklearn/decomposition/_dict_learning.py

```diff
@@ -1503,19 +1504,15 @@ def transform(self, X, y=None):
     def _more_tags(self):
         return {
             "requires_fit": False,
-            "preserves_dtype": [np.float64, np.float32],
+            "preserves_dtype": [np.float64],
+            "stateless": True,
         }
 
     @property
     def n_components_(self):
         """Number of atoms."""
         return self.dictionary.shape[0]
 
-    @property
-    def n_features_in_(self):
-        """Number of features seen during `fit`."""
-        return self.dictionary.shape[1]
```

Comment:
Couldn't we still expose `n_features_in_` when `getattr(self, "dictionary", None) is not None`?

This would be less disruptive and would not require to change `test_sparse_coder_n_features_in` (also both cases could be tested instead: with and without calling `fit`).

Reply:
Indeed, I can easily preserve the behaviour with:

```python
    @property
    def n_features_in_(self):
        """Number of features seen during :term:`fit`."""
        return getattr(self, "_n_features_in_", self.dictionary.shape[1])

    @n_features_in_.setter
    def n_features_in_(self, value):
        self._n_features_in_ = value
```

However, it will break our API and more precisely the `check_n_features_in` since `n_features_in_` should not be set if `fit` is not called.

I don't know what is best here.

Reply:
I think it makes more sense for `fit` to set `n_features_in_`, and that would also satisfy our API test, i.e., no property kinda thing.